File: nodejs/packages/node-red-contrib-periph-gnss/nodes/neo6/neo6.js
'use strict';

module.exports = function(RED) {
    const { UARTConnection } = require('periph/src/connection/uart');
    const { I2CConnection }  = require('periph/src/connection/i2c');
    const { SPIConnection }  = require('periph/src/connection/spi');
    const { NEO6Full }      = require('periph/src/chips/gnss/neo6');

    function NEO6Node(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        node.busType = config.busType || 'uart';
        node.ready = false;

        (async () => {
            try {
                let connection;
                if (node.busType === 'i2c') {
                    const addr = parseInt(config.addr) || 0x42;
                    connection = new I2CConnection(parseInt(config.bus) || 1, addr);
                } else if (node.busType === 'spi') {
                    connection = new SPIConnection(parseInt(config.bus) || 0, parseInt(config.device) || 0,
                        { mode: 0, maxSpeedHz: 200_000 });
                } else {
                    connection = new UARTConnection(config.port || '/dev/ttyS0',
                        { baudRate: parseInt(config.baud) || 9600 });
                    await connection.open();
                }
                node.connection = connection;
                node.driver = new NEO6Full(connection, node.busType);
                node.ready = true;
            } catch (e) {
                node.error('NEO-6 init failed: ' + e.message);
            }
        })();

        node.on('input', async function(msg, send, done) {
            if (!node.ready || !node.driver) {
                node.error('NEO-6 not ready', msg);
                done();
                return;
            }
            try {
                // A GPS burst (~7 sentences at the module's 1 Hz output rate)
                // is far more than one byte; drain what's currently buffered
                // each time this node fires rather than reading a single
                // byte. Time-boxed rather than a fixed iteration count: on
                // UART a byte-less update() only resolves after the
                // connection's read timeout, so an unbounded iteration count
                // could block this node for a very long time once the
                // stream goes quiet between bursts.
                const deadline = Date.now() + 900;
                while (Date.now() < deadline) {
                    await node.driver.update();
                }
                const d = node.driver;
                msg.payload = {
                    lat: d.latitude(),
                    lon: d.longitude(),
                    alt: d.altitude(),
                    fix: d.fix(),
                    satellites: d.satellites(),
                    speed: d.speed(),
                    course: d.course(),
                    time: d.utcTime(),
                    date: d.utcDate(),
                    hdop: d.hdop()
                };
                send(msg);
                done();
            } catch (e) {
                done(e);
            }
        });

        node.on('close', async function(done) {
            if (node.connection) {
                try { await node.connection.close(); } catch (_) { /* already closed */ }
            }
            done();
        });
    }

    RED.nodes.registerType('periph-neo6', NEO6Node);
};

File: python/uiflow2/gnss/neo6/NEO6.py
"""
file     NEO6
time     2026-08-27
author
email
license  Apache License 2.0
"""

from periph.connection.uart_auto import UARTConnection
from periph.chips.gnss.neo6 import NEO6Full


class NEO6:
    """
    note:
        en: ''
    details:
        color: '#C084FC'
        link: https://github.com/tuhde/Periph
        image: ''
        category: Custom
    example: ''
    """

    def __init__(self, port: int = 1, baudrate: int = 9600, tx: int = 0, rx: int = 0):
        """
        label:
            en: '%1 init port %2 baudrate %3 tx %4 rx %5'
        params:
            port:
                name: port
                type: int
                default: '1'
                field: number
            baudrate:
                name: baudrate
                type: int
                default: '9600'
                field: number
            tx:
                name: tx
                type: int
                default: '0'
                field: number
            rx:
                name: rx
                type: int
                default: '0'
                field: number
        """
        connection = UARTConnection(port=port, baudrate=baudrate, tx=tx, rx=rx)
        self._driver = NEO6Full(connection)

    def update(self) -> bool:
        """
        label:
            en: '%1 update (got new fix?)'
        """
        return self._driver.update()

    def latitude(self) -> float:
        """
        label:
            en: '%1 latitude (°)'
        """
        return self._driver.latitude()

    def longitude(self) -> float:
        """
        label:
            en: '%1 longitude (°)'
        """
        return self._driver.longitude()

    def altitude(self) -> float:
        """
        label:
            en: '%1 altitude (m)'
        """
        return self._driver.altitude()
